src/accountsdb/buffer_pool.zig

```diff
@@ -0,0 +1,1174 @@
+            if (n_submitted != sent_reads) unreachable; // did something else submit an event?
+
+            // would be nice to get rid of this alloc
+            const cqes = try allocator.alloc(std.os.linux.io_uring_cqe, n_submitted);
+            defer allocator.free(cqes);
+
+            // check our completions in order to set the frame's size;
+            // we need to wait for completion to get the bytes read
+            const cqe_count = try self.io_uring.copy_cqes(cqes, n_submitted);
+            if (cqe_count != n_submitted) unreachable; // why did we not receive them all?
+            for (0.., cqes) |i, cqe| {
+                if (cqe.err() != .SUCCESS) {
+                    std.debug.panic("cqe err: {}, i: {}", .{ cqe, i });
+                }
+                const f_idx = cqe.user_data;
+                const bytes_read: FrameOffset = @intCast(cqe.res);
+                if (bytes_read > FRAME_SIZE) unreachable;
+
+                // TODO: atomics
+                self.frames_metadata.size[f_idx] = bytes_read;
+            }
+        }
+
+        return CachedRead{
+            .buffer_pool = self,
+            .frame_indices = frame_indices,
+            .first_frame_start_offset = @intCast(file_offset_start % FRAME_SIZE),
+            .last_frame_end_offset = @intCast(((file_offset_end - 1) % FRAME_SIZE) + 1),
+        };
+    }
+
+    fn readBlocking(
+        self: *BufferPool,
+        /// used for temp allocations, and the returned .indices slice
+        allocator: std.mem.Allocator,
+        file: std.fs.File,
+        file_id: FileId,
+        /// inclusive
+        file_offset_start: FileOffset,
+        /// exclusive
+        file_offset_end: FileOffset,
+    ) (error{
+        InvalidArgument,
+        OutOfMemory,
+        InvalidKey,
+        CannotResetAlive,
+        CannotOverwriteAliveInfo,
+    } || std.posix.PReadError)!CachedRead {
+        const frame_indices = try self.computeFrameIndices(
+            file_id,
+            allocator,
+            file_offset_start,
+            file_offset_end,
+        );
+        errdefer allocator.free(frame_indices);
+
+        // update found frames in the LFU (we don't want to evict these in the next loop)
+        for (frame_indices) |f_idx| {
+            if (f_idx == INVALID_FRAME) continue;
+            try self.eviction_lfu.insert(self.frames_metadata, f_idx);
+            if (!self.frames_metadata.rc[f_idx].acquire()) {
+                // frame has no handles, but memory is still valid
+                self.frames_metadata.rc[f_idx].reset();
+            }
+        }
+
+        // fill in invalid frames with file data, replacing invalid frames with
+        // fresh ones.
+        for (0.., frame_indices) |i, *f_idx| {
+            if (f_idx.* != INVALID_FRAME) continue;
+            // INVALID_FRAME => not found, read fresh and populate
+
+            const frame_aligned_file_offset: FileOffset = @intCast((i * FRAME_SIZE) +
```

Comment:
any reason thse are two seperate loops vs one loop and two if branches ? less continues the better

Reply:
We want to iterate over found frames first, as if we don't, the next loop could cause them to be evicted from the hierarchical fifo. Having alive rcs protects them from eviction in the next loop (& the _lfu.insert means we don't waste time trying to evict them). Finding a frame means we want to use it, whereas eviction means overwriting a frame; we could get two of the same frame indices for different parts of the read (which is very bad!)